core/controllers/load_functions.py

```diff
@@ -0,0 +1,257 @@
+    # If this org already in DB, merge to re-use pk, otherwise use add_all as per loop continuation
+    if organisation_exists:
+        org_to_merge = models[0]
+        org_to_merge.id = organisation_exists.id
+        db.session.merge(org_to_merge)  # There can only be 1 org per ingest.
+    else:
+        db.session.add_all(models)
+
+
+def load_outputs_outcomes_ref(workbook: dict[str, pd.DataFrame], mapping: DataMapping):
+    """
+    Loads data into the 'Outputs_Ref' or 'Outcomes_Ref' tables.
+
+    The function first retrieves the relevant data from the workbook using the provided mapping.
+    It then processes the data to get outcomes and outputs to insert,
+    and finally adds all the models to the database session.
+
+    :param workbook: a dictionary of DataFrames of table data to be inserted into the db.
+    :param mapping: the mapping of the relevant DataFrame to its attributes as they appear in the db.
+    """
+    worksheet = workbook[mapping.table]
+    models = mapping.map_data_to_models(worksheet)
+
+    models = get_outcomes_outputs_to_insert(mapping, models)
+    db.session.add_all(models)
+
+
+def generic_load(workbook: dict[str, pd.DataFrame], mapping: DataMapping, submission_id):
+    """
+    Generic function to load data into a specified table.
+
+    The function takes a workbook, a mapping of relevant data, and a submission ID.
+    It adds the submission ID to the worksheet, maps the data to models using the provided mapping,
+    and adds all the models to the database session.
+
+    :param workbook: a dictionary of DataFrames of table data to be inserted into the db.
+    :param mapping: the mapping of the relevant DataFrame to its attributes as they appear in the db.
+    :param submission_id: the ID of the submission associated with the data.
+    """
+    worksheet = workbook[mapping.table]
+    worksheet["Submission ID"] = submission_id
+    models = mapping.map_data_to_models(worksheet)
+
+    db.session.add_all(models)
+
+
+def get_submission_id(workbook: dict[str, pd.DataFrame]) -> str:
+    """
+    Retrieves or generates a submission ID based on the information in the provided workbook.
+
+    The function extracts the reporting round and programme ID from the workbook.
+    If a matching programme exists in the same reporting round, it finds the corresponding project and submission.
+    The submission ID is then retrieved, and the existing submission is removed.
+    If there is no matching programme for the current round, a new submission ID is generated.
+
+    :param workbook: a dictionary of DataFrames containing table data to determine the submission ID.
+    :return: a string representing the submission ID.
+    """
+    reporting_round = int(workbook["Submission_Ref"]["Reporting Round"].iloc[0])
+    programme_id = workbook["Programme_Ref"]["Programme ID"].iloc[0]
+    # if already added this round, this entity used to drop existing round data
+    programme_exists_same_round = (
```

Comment:
as per other comment, could consider splitting DB queries into separate functions that live in queries module?